sale_delivery_state/models/sale_order.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2018 Akretion (http://www.akretion.com).
+# @author Pierrick BRUN <pierrick.brun@akretion.com>
+# Copyright 2018 Camptocamp
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, api
+from odoo.tools import float_is_zero, float_compare
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    delivery_state = fields.Selection([
+        ('no', 'No delivery'),
+        ('unprocessed', 'Unprocessed'),
+        ('partially', 'Partially processed'),
+        ('done', 'Done')],
+        string='Delivery state',
+        compute='_compute_delivery_state',
+        store=True
+    )
+
+    force_delivery_state = fields.Boolean(
+        string='Force delivery state',
+        help=("Allow to enforce done state of delivery, for instance if some"
+              " quantities were cancelled")
+    )
+
+    def _all_qty_delivered(self):
+        """
+        Returns True if all line have qty_delivered >= to ordered quantities
+
+        If `delivery` module is installed, ignores the lines with delivery costs
+
+        :returns: boolean
+        """
+        self.ensure_one()
+        # Skip delivery costs lines
+        sale_lines = self.order_line.filtered(lambda rec: not rec._is_delivery())
```

Comment:
Shoudn't we exclude all service products instead on only delivery service product ?

Reply:
No, service can be delivered manually by setting the qty on the sale.order.line.
This module is generic and can be used with or without delivery module.
In case that you do not have "stockable" product than the state delivered mean that you have finish your job.

Reply:
Ok for me